server/datastore/mysql/mdm.go

```diff
@@ -1038,23 +1038,35 @@ func (ds *Datastore) GetHostCertAssociationsToExpire(ctx context.Context, expiry
 	//
 	// Note that we use GROUP BY because we can't guarantee unique entries
 	// based on uuid in the hosts table.
-	stmt, args, err := sqlx.In(
-		`SELECT
-			h.uuid as host_uuid,
-			ncaa.sha256 as sha256,
-			COALESCE(MAX(hm.fleet_enroll_ref), '') as enroll_reference
-		 FROM
-			nano_cert_auth_associations ncaa
-			JOIN hosts h ON h.uuid = ncaa.id
-			LEFT JOIN host_mdm hm ON hm.host_id = h.id
-		 WHERE
-			cert_not_valid_after BETWEEN '0000-00-00' AND DATE_ADD(CURDATE(), INTERVAL ? DAY)
-			AND renew_command_uuid IS NULL
-		GROUP BY
-			host_uuid, ncaa.sha256, cert_not_valid_after
-		ORDER BY cert_not_valid_after ASC
-		LIMIT ?
-		`, expiryDays, limit)
+	stmt, args, err := sqlx.In(`
+SELECT
+    h.uuid AS host_uuid,
+    ncaa.sha256 AS sha256,
+    COALESCE(MAX(hm.fleet_enroll_ref), '') AS enroll_reference
+FROM
+    nano_cert_auth_associations ncaa
+JOIN
+    hosts h ON h.uuid = ncaa.id
+LEFT JOIN
+    host_mdm hm ON hm.host_id = h.id
+-- grab only the latest certificate associated with this device 
+JOIN (
+    SELECT
+        id,
+        MAX(created_at) AS latest_created_at
+    FROM
+        nano_cert_auth_associations
+    GROUP BY
+        id
+) latest_ncaa ON ncaa.id = latest_ncaa.id AND ncaa.created_at = latest_ncaa.latest_created_at
```

Comment:
I think that technically this could still result in multiple SCEP entries being returned for a given host if there are two `nano_cert_auth_associations` entries for the same host and same timestamp (with mysql's poor default resolution of 1 second). That's probably very unlikely though?

Reply:
I found this exact scenario during testing. I also thought it was very unlikely, but honestly I don't know 🤔 

what should we do for that case? one good and easy thing to do might be to increase the timestamp resolution of that column, another easy win is to add `DISTINCT` somewhere

there's also a safeguard in the cron now, juuust in case.

Reply:
Increasing the resolution just makes it less likely to happen though. I think there's a query to make it impossible, but it's a bit hard to figure out. If the "latest entry" table groups by `id` and sorts by `created_at DESC, sha256` with a `LIMIT 1`? I'm not quite sure how to fit it in the greater query, but I feel like something down this path would do it? (maybe it has to be in the WHERE clause and not in the FROM/JOIN) I'll try something in the REPL and post if I make any progress. But your PR is probably already a huge benefit as-is, just not for this edge-case.

Reply:
Simplified query, but I think something like that?

```
select id, sha256    from nano_cert_auth_associations n1   where sha256 = (select sha256 from nano_cert_auth_associations n2 WHERE n1.id = n2.id ORDER
BY n2.created_at DESC, sha256 ASC LIMIT 1);
```

Reply:
(at this point it arbitrarily chooses one at the latest timestamp, don't know if that matters which one gets used?)

Reply:
agreed, I was thinking that while not completely avoiding it, having a better resolution might be a good enough solution.

Thank you! I'm also trying a few things out to see if I find a good one.

Reply:
giving that a try, thanks! I don't think it matters at all

Reply:
@mna amazing idea, can you give a sanity check to the diff and let me know what you think? https://github.com/fleetdm/fleet/pull/19313/commits/7c587c2456b891f6014d8cf88122087d83e86b82